File: fibonacci/fib.scala
import scala.collection.mutable.HashMap 

object fib{

    // plain recursion
    // (O) = 2^n

    def fibPlain(n: Int): Integer = {
        if (n <= 2) {1}
        else {fibPlain(n-1) + fibPlain(n-2)}
    }

    //memoization 
    // (O) = n
    
    def fibMemo(n: Long, memo: HashMap[Long, Long]): Long = {
        if (memo.contains(n)) {
            memo.getOrElse(n,0)
        } else {
            if( n <= 2) {1}
            else {            
                memo(n) = fibMemo(n-1, memo) + fibMemo(n-2, memo)
                memo.getOrElse(n, 0)
            } 
        }
    }

    def main(args: Array[String]): Unit = {
        println(fibPlain(10))
        var memo: HashMap[Long, Long] = new HashMap[Long, Long]()
        println(fibMemo(50, memo))
    }
}

File: fibonacci/fib.java
package fibonacci;

import java.util.HashMap;

class fib {

    // plain recursion
    // (O) = 2^n

    public static int fibPlain(int n) {
        if (n <= 2) {
            return 1;
        } else {
            return fibPlain(n - 1) + fibPlain(n - 2);
        }
    }
    
    // memoization
    // (O) = n
    
    public static Long fibMemo(Long n, HashMap<Long, Long> memo) {
        if (memo.containsKey(n)) {
            return memo.get(n);
        } else {
            if (n <= 2) {
                return (long) 1;
            } else {
                memo.put(n, fibMemo(n - 1, memo) + fibMemo(n - 2, memo));
                return memo.get(n);
            }
        }
    }
    public static void main(String args[]) {
        System.out.println(fibPlain(10));
        HashMap<Long, Long> memo = new HashMap<Long, Long>();
        System.out.println(fibMemo((long) 50, memo));
    }
}